TypeScript support in core

sneridagh opened the pull request:
So, let's put this on the table and discuss it during the Beethoven Sprint.

To be clear: I'm (still) far from being sold on TypeScript, and have profound doubts on how it could benefit Volto in short/mid term (as I always did). For a number of reasons:

- Plone/Volto developers base are not familiar with it,
- let alone old Plone community members who wants to adopt Volto for their next projects.
- The community lacks of "TypeScript Wizards(TM)" to rely on when things get weary.
- Problems derived from adding an additional abstraction layer to the equation outweighs all the possible benefits that it could bring, and a whole new category of errors (typings) to solve, instead of focusing in features and getting things done.
- "Type safety" is still a concept that, personally speaking, is overrated and often misunderstood by "hurray, my IDE autocompletes everything". Unit, acceptance testing and documentation are as important as "type safety".

_However_, once named the "cons", there are also some "pros". 

- Best way to document code internally, signatures, props, config object, etc.
- Could be the entry point and the enabler of some other complex things.
- The whole ecosystem is migrating to it, and the typings permeate the IDEs and your code.
- Volto is becoming stable enough to start introducing it.

Summarising, the proposal is *not* to migrate Volto to TypeScript, but having the ability to use it here and there in core if appropriate in some specific areas to gain knowledge and confidence as a tool, and artefact, not as a must.

I know could be a drawback and might be chaotic to have some things in JS and others in TS. As said, let's discuss it during the sprint, and see how it goes.

/cc @mtoepfl I'd like to know your feedback about the PR, if you please could take a look, it would be great!

Regarding implementation details, I added a `tsconfig.json`, adding TypeScript support for `.ts` files only as Razzle already supports it. In the ESlint side, I replaced the (deprecated) `babel-eslint` with `@babel/eslint-parser`, and use the TypeScript ESlint parser only on the `.ts` files. So effectively this way Volto supports both JS and TS at the same time.

*UPDATE*: I fixed some ESlint issues that the new parser and rules was complaining about. The important changes related to this PR are in the root.

*UPDATE (2023-04-17)*: I've managed to make the change non-breaking for 16 series: https://github.com/plone/volto/pull/4705 Undoubtedly, there could be backport issues still, but it's doable.

pnicolli commented:
😃 

> * Plone/Volto developers base are not familiar with it,

Agreed that this is an issue, it is my main issue on this topic, the other ones are just smaller obstacles.

> * let alone old Plone community members who wants to adopt Volto for their next projects.

All of these people can use JavaScript in their projects as they would today. Adopting TS in core does not mean forcing people to adopt it for projects that use Volto. It would just improve their developer experience actually.

> * The community lacks of "TypeScript Wizards(TM)" to rely on when things get weary.

Well, we kind of often lacked in JavaScript experience in the community, and Volto is a simple enough case in my opinion for TypeScript not become a enormous problem. Also, typings can be rough and improved later, for possible complex situations.

> * Problems derived from adding an additional abstraction layer to the equation outweighs all the possible benefits that it could bring, and a whole new category of errors (typings) to solve, instead of focusing in features and getting things done.

I believe you might be underestimating the benefits here, with blocks, config and api being the most complex things to work with and the ones that would benefit the most from TS.

> * "Type safety" is still a concept that, personally speaking, is overrated and often misunderstood by "hurray, my IDE autocompletes everything". Unit, acceptance testing and documentation are as important as "type safety".

Yes, those things are crucial, TypeScript aims at being an addition to those, it is not there to allow us to remove tests or docs.
My preferred TS benefit is consistency in refactoring, something that has been done several times in this product.

> _However_, once named the "cons", there are also some "pros".
> 
> * Best way to document code internally, signatures, props, config object, etc.
> * Could be the entry point and the enabler of some other complex things.

Zod? 😃 

> * The whole ecosystem is migrating to it, and the typings permeate the IDEs and your code.
> * Volto is becoming stable enough to start introducing it.

Actually it would have been more beneficial before being stable than after being stable, for the reasons explained above. But it's never too late, and Volto will never at perfect stability.

> Summarising, the proposal is _not_ to migrate Volto to TypeScript, but having the ability to use it here and there in core if appropriate in some specific areas to gain knowledge and confidence as a tool, and artefact, not as a must.

Imho, the goal should be to enable TS, yes, but with the long term dream to completely migrate, otherwise the benefits are cut in half (but still there).

> I know could be a drawback and might be chaotic to have some things in JS and others in TS. As said, let's discuss it during the sprint, and see how it goes.
> 
> Regarding implementation details, I added a `tsconfig.json`, adding TypeScript support for `.ts` files only as Razzle already supports it. In the ESlint side, I replaced the (deprecated) `babel-eslint` with `@babel/eslint-parser`, and use the TypeScript ESlint parser only on the `.ts` files. So effectively this way Volto supports both JS and TS at the same time.

iirc this is basically the same we suggest in the docs, so it's perfectly fine by me

mtoepfl commented:
Hey @sneridagh. Thanks for mentioning. I'm happy to read this message this evening.

I'm happy to revisit the changes I did on the sprints at the conference and push it to this branch.

I only had some problems with some `export * from 'xxxx';` statements, but the rest looks really promising.

A problem is that you have to touch many files which are hard to merge if the modified files are changed by other pull requests. So a discussion/decision beforehand could save a lot of time.

I would love to answer questions on typings. Mostly it's very straight forward and easy to learn for app-parts. For some tweeks in "library"-parts, I'm sure, that there are many real "TypeScript Wizard" on the internet who will be happy to jump in and give feedback if needed.

I am interested to see the comments of the other core developers.

mtoepfl commented:
@pnicolli:

> My preferred TS benefit is consistency in refactoring, something

100% dito: eg. thinking of refactoring the fetch/redux logic

+1 for this and all other points you mentioned

sneridagh commented:
@mtoepfl I knew you'd like it ;) I'd love to have the work you did during the last PC sprint!

Btw, could we have it separated from this PR? So we can keep this one as the PR that enables the support, then all the other related work.

I'd like also to have full ok and support from the Volto Team and hear all the feedback about it before proceeding.

Regarding the `export X from Y` problem, it's fixed.

sneridagh commented:
@tiberiuichim if you have a moment, due to upgrades to the ESlint plugins, now react-hooks plugin is complaining about:

```
/Users/sneridagh/Development/plone/volto/src/components/manage/Blocks/Search/hocs/withQueryString.jsx
  18:11  warning  The 'indexes' logical expression could make the dependencies of useEffect Hook (at line 24) change on every render. Move it inside the useEffect callback. Alternatively, wrap the initialization of 'indexes' in its own useMemo() Hook  react-hooks/exhaustive-deps

✖ 1 problem (0 errors, 1 warning)
```

https://github.com/plone/volto/pull/4662/files#diff-fc8bdb4141e8455cee5f70633025f8027f1449a064b16bf87b3530d28ee2c0b2L18

Could you please take a look? Thanks!

sneridagh commented:
@plone/volto-team I just realised of one thing.

Since I had to update ESlint to the latests version and I changed the parser to use `@babel/eslint-parser`, and I've updated all the ESlint plugins, it must be a breaking one (since now it reports more things than before, affecting all projects out there.

The CI is green for a project (but we don't have any .ts component yet), so we don't know how this will affect them. We'll have to see, but I expect some adjustments there too (at least jsconfig->tsconfig) changes.

Marking it as breaking change shouldn't be any problem at first sight, but that means that we won't be able to port _anything_ TS based to 16... which is an important drawback...

Something to talk about during the sprint.

sneridagh commented:
I've managed to make the change non-breaking for 16 series: https://github.com/plone/volto/pull/4705 Undoubtedly, there could be backport issues still, but it's doable.

stevepiercy reviewed (changes requested):
We need docs for this feature addition. Something like:

```{versionadded} 16.x
We got TypeScript! w00t! 🥳 🎈 

But you don't *have* to use it, you knuckle-dragging mouth breathing luddites.
```

Oh, OK, maybe not quite like that (which would be worse than no docs), but I'm sure y'all can improve upon it. 😜

pnicolli commented on tsconfig.json:
```diff
@@ -0,0 +1,33 @@
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": false,
```
I would sooooooo want to start this with `true`. Does it break stuff?

sneridagh replied:
Wish granted.

pnicolli reviewed (approved):
Just a small comment, but LGTM.

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+### TypeScript support in Volto
+
+```{versionadded} 17.0.0-alpha.22
+We have now full support of TypeScript in Volto core.
````
```suggestion
We added full support of TypeScript in Volto core.
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+```{versionadded} 17.0.0-alpha.22
+We have now full support of TypeScript in Volto core.
+No existing code has been migrated.
+You still have the freedom to develop using JavaScript, it's your choice if you want to do it in TypeScript.
````
```suggestion
You still have the freedom to develop using JavaScript.
It's your choice whether you want to do it in TypeScript.
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
```diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+No existing code has been migrated.
+You still have the freedom to develop using JavaScript, it's your choice if you want to do it in TypeScript.
+In the past, you could add TypeScript support for Volto add-ons.
+It's also available for projects too now if you upgrade dependencies (see next section).
```
```suggestion
It's also available for projects now, if you upgrade dependencies (see next section).
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+It's also available for projects too now if you upgrade dependencies (see next section).
+```
+
+You should update your projects if you want to support TypeScript in your projects as well.
````
```suggestion
If you want to support TypeScript in your projects, you should update your projects.
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
```diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+
+You should update your projects if you want to support TypeScript in your projects as well.
+
+Changes in your `package.json`:
```
```suggestion
Make the following changes in your {file}`package.json`:
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
```diff
@@ -52,6 +52,40 @@ or use Webpack plugins, you might need to make adjustments.
+ESlint uses a library to parse the language under analysis.
+The one used was long deprecated and didn't supported both TypeScript and JavaScript.
+We upgraded the ESlint parser to use `@babel/eslint-parser`.
+This means that could be that when you upgrade your projects, some new violations appear in your project.
```
```suggestion
This means when you upgrade your project, some new violations may appear.
```

stevepiercy reviewed (changes requested):
Minor English and MyST syntax, grammar, and style.

sneridagh commented:
*Decision Volto Team Meeting (2023-09-12):* It will be included in Volto 17. However, it needs to be documented properly. @sneridagh commits himself to work on this during the PloneConf Sprint (and other documentation matters).

- How to use React TypeScript
- How to configure IDEs to support TypeScript
- Which are the limits of usage (if any) and the expectations (eg. no code would be migrated, it won't be mandatory for any code).
- How to handle backports to stable (if any).

mtoepfl commented:
🎉 I'm going to join you at coming PloneConf Sprint. 

I could try to cherry pick my changes of the PloneConf Sprint last year. Before I planned to write a benchmark for creating brains with many fields in metadata for ZCatalog. I feel it's getting over proportional slower in reading, when adding many fields to metadata. Maybe there ist a bottleneck. But finding this out should not take too long.

Is it possible to merge some of the refactoring pull requests to functional components before the Sprint?

sneridagh commented:
@mtoepfl awesome! I might multiplex this and docs this year. 

Could be that your changes overlap with this PR I made: https://github.com/plone/volto/pull/4949

We will have to merge them both and you review my crappy TS ;) 

I can't wait!

sneridagh commented:
@stevepiercy I made some additions to the docs, could you please take a look?

@pnicolli I tried this in existing projects and seems to work fine, even without changing the jsconfig->tsconfig so I guess we are good.

I updated the docs to reflect all that should be done in order to update a project to TS (which is acceptable).

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+It's your choice whether you want to do it in TypeScript.
+In the past, you could add TypeScript support for Volto add-ons.
+It's also available for projects now, if you upgrade dependencies (see next section).
+```
````
Let's simplify this block, moving most content outside of the block.

````suggestion
```{versionadded} 17.0.0-alpha.22
```

We added full support of TypeScript in Volto core.

No existing code has been migrated.
You may write code in either JavaScript or TypeScript.
The choice is yours.

Previously developers had the option to support TypeScript only in their Volto add-ons.
Now it's available in Volto projects, as long as you upgrade your project dependencies.
````

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+It's also available for projects now, if you upgrade dependencies (see next section).
+```
+
+If you want to support TypeScript in your projects, you should update your projects.
````
```suggestion
To support TypeScript in your projects, you must update your project as follows.
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
```diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+
+If you want to support TypeScript in your projects, you should update your projects.
+
+Make the following changes in your {file}`package.json`:
```
```suggestion
Edit your {file}`package.json`:
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+```
+
+```{note}
+Could be that after this, you experience hoisting problems and some packages can't be found on start.
````
```suggestion
After making this change, you might experience hoisting problems and some packages can't be found on start.
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+In that case, make sure you reset your `yarn.lock` by deleting it and start with a clean environment.
+```
+
+In case that you want to use TypeScript in your projects, you'll need to introduce a `tsconfig.json` file in it, and remove the existing `jsconfig.json`. You can use the one provided by de generator as template, or use your own:
````
```suggestion
To use TypeScript in your projects, you'll need to introduce a TypeScript configuration file {file}`tsconfig.json`, and remove the existing file {file}`jsconfig.json`.
You can use the one provided by the generator as a template, or use your own:
```

pnicolli replied:
```suggestion
In case that you want to use TypeScript in your projects, you'll need to introduce a `tsconfig.json` file in it, and remove the existing `jsconfig.json`. You can use the one provided by the generator as template, or use your own:
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+}
+```
+
+If you are using `mrs-developer` in your project, update the command in `Makefile`:
````
```suggestion
If you use `mrs-developer` in your project, update the command in {file}`Makefile`:
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+```diff
+--- a/Makefile
++++ b/Makefile
+@@ -59,7 +59,8 @@ preinstall: ## Preinstall task, checks if missdev (mrs-developer) is present and
````
"and" ...what else? Don't leave me hanging!

It should match the Makefile.

```suggestion
preinstall: ## Preinstall task, checks if missdev (mrs-developer) is present and runs it
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
`````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+```
+
+````{note}
+Do not forget to run `mrs-developer` after doing it, so the configuration gets in the right place (`tsconfig.json`) now.
`````
```suggestion
After editing your {file}`Makefile`, run `mrs-developer` with the following command, so the configuration gets in the right place ({file}`tsconfig.json`).
```

stevepiercy commented on docs/source/upgrade-guide/index.md:
`````diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+
+````{note}
+Do not forget to run `mrs-developer` after doing it, so the configuration gets in the right place (`tsconfig.json`) now.
+Run
`````
```suggestion

```

stevepiercy commented on docs/source/upgrade-guide/index.md:
```diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+
+### Upgrade ESlint and use `@babel/eslint-parser`
+
+ESlint uses a library to parse the language under analysis.
```
````suggestion
```{versionchanged} 17.0.0-alpha.22
```
ESlint uses a library to parse the language under analysis.
````

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
 
 # TypeScript
 
+TypeScript support (5.2.2) has been added to Volto core.
```
I don't think we want to specify a version of third party software that will need to be updated with each release. I am lazy. 😉 

````suggestion
```{versionadded} 17.0.0-alpha.22
```
TypeScript support has been added to Volto core.
````

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
 
+TypeScript support (5.2.2) has been added to Volto core.
+ESlint has been configured in order to accept code base written both in JavaScript and in TypeScript.
+The usage of TypeScript in projects and add-ons was possible already since Volto 14 and it's added as default in the project generator since Volto 17.
```
```suggestion
TypeScript in projects and add-ons has been supported since Volto 14.
It is now supported in the project generator by default.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+## TypeScript in core
+
+The internal policy will be as follows:
```
```suggestion
The internal policy for TypeScript support in Volto core will be as follows.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+The internal policy will be as follows:
+
+- No existing code will be migrated at this point (eg. Volto 17.0.0 won't have any TypeScript present)
```
```suggestion
-   No existing code has been migrated.
    Volto 17.0.0 has no TypeScript code present.
    Only *support* for TypeScript has been added.
```

pnicolli commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+- No existing code will be migrated at this point (eg. Volto 17.0.0 won't have any TypeScript present)
+- Code in core can be written in both JavaScript and TypeScript
+- In any case you are forced to migrate existing code to TypeScript
```
I am not sure I understand this bullet point here

stevepiercy replied:
I think you meant *not*, yes?

```suggestion
-   You are not forced to migrate existing code to TypeScript.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+The internal policy will be as follows:
+
+- No existing code will be migrated at this point (eg. Volto 17.0.0 won't have any TypeScript present)
+- Code in core can be written in both JavaScript and TypeScript
```
```suggestion
-   Code in core can be written in either JavaScript or TypeScript, or both.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+- No existing code will be migrated at this point (eg. Volto 17.0.0 won't have any TypeScript present)
+- Code in core can be written in both JavaScript and TypeScript
+- In any case you are forced to migrate existing code to TypeScript
+- It's up to the developer to choose which one to use
```
```suggestion
    It's your choice which one to use.
```

pnicolli commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
 
-It is possible to enable TypeScript support when developing you project with Volto.
-Volto itself is not written in TypeScript, but it is ready to use it if you want.
+In case that you want to use TypeScript in your projects, you'll need to introduce a `tsconfig.json` file in it, and remove the existing `jsconfig.json`. You can use the one provided by de generator as template, or use your own:
```
```suggestion
In case that you want to use TypeScript in your projects, you'll need to introduce a `tsconfig.json` file in it, and remove the existing `jsconfig.json`. You can use the one provided by the generator as template, or use your own:
```

stevepiercy replied:
```suggestion
To use TypeScript in your projects, you'll need to introduce a TypeScript configuration file {file}`tsconfig.json`, and remove the existing file {file}`jsconfig.json`.
You can use the one provided by the generator as a template, or use your own:
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+- Code in core can be written in both JavaScript and TypeScript
+- In any case you are forced to migrate existing code to TypeScript
+- It's up to the developer to choose which one to use
+- In all cases use common sense and be aware of code that could potentially have to be backported
```
```suggestion
-   If you need to backport code to a version of Volto before 17.0.0, you must make the code compatible without TypeScript.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+- In any case you are forced to migrate existing code to TypeScript
+- It's up to the developer to choose which one to use
+- In all cases use common sense and be aware of code that could potentially have to be backported
+- If the developer chooses TypeScript and the feature has to be backported, it's up to the developer to backport it to non-TypeScript version in legacy (eg. LTS) versions
```
Redundant to previous statement.

```suggestion
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+## TypeScript learning resources
+
+[TypeScript Official Handbook](https://www.typescriptlang.org/docs/handbook/intro.html)
```
```suggestion
-   [TypeScript Official Handbook](https://www.typescriptlang.org/docs/handbook/intro.html)
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+[TypeScript Official Handbook](https://www.typescriptlang.org/docs/handbook/intro.html)
+
+[TotalTypeScript React with TypeScript Free course](https://www.totaltypescript.com/tutorials/react-with-typescript)
```
```suggestion
-   [TotalTypeScript React with TypeScript Free course](https://www.totaltypescript.com/tutorials/react-with-typescript)
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+[TotalTypeScript React with TypeScript Free course](https://www.totaltypescript.com/tutorials/react-with-typescript)
+
+[TotalTypeScript Beginner's TypeScript Free course](https://www.totaltypescript.com/tutorials/beginners-typescript)
```
```suggestion
-   [TotalTypeScript Beginner's TypeScript Free course](https://www.totaltypescript.com/tutorials/beginners-typescript)
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+## TypeScript in projects and add-ons
+
+It is possible to enable TypeScript support when developing you project with Volto.
```
```suggestion
You can enable TypeScript support when developing your project in Volto.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+## TypeScript in projects and add-ons
+
+It is possible to enable TypeScript support when developing you project with Volto.
+If you are using the Volto generator, this is the default, and you can already start using TypeScript in it.
```
```suggestion
If you use the Volto generator, TypeScript is enabled by default, and you can already start using TypeScript in it.
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+### Enable it in your Volto 17 project
+
+If you are already using Volto 17 in your projects, using a boilerplate generated before Volto 17.0.0-alpha.27 and the generator 7.0.0-alpha.7 you can follow the following instructions:
```
```suggestion
If you already use Volto 17 in your projects, using a boilerplate generated before Volto 17.0.0-alpha.27 and the generator 7.0.0-alpha.7, you can enable TypeScript support as follows:
```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+
+If you are already using Volto 17 in your projects, using a boilerplate generated before Volto 17.0.0-alpha.27 and the generator 7.0.0-alpha.7 you can follow the following instructions:
+
+Make the following changes in your {file}`package.json`:
```
```suggestion
Edit your {file}`package.json`:
```

stevepiercy commented on docs/source/contributing/typescript.md:
````diff
@@ -9,14 +9,126 @@ myst:
+
 ```{note}
-This feature is available since Volto 14. Please note that TypeScript is *not* allowed in Volto core, that's why it's not enabled in the core by default. This feature enables the use of TypeScript (as opt-in) in projects and in add-ons.
+Could be that after this, you experience hoisting problems and some packages can't be found on start.
````
```suggestion
After making this change, you might experience hoisting problems and some packages can't be found on start.
```

stevepiercy commented on docs/source/contributing/typescript.md:
````diff
@@ -9,14 +9,126 @@ myst:
+}
+```
+
+If you are using `mrs-developer` in your project, update the command in `Makefile`:
````
```suggestion
If you use `mrs-developer` in your project, update the command in {file}`Makefile`:
```

stevepiercy commented on docs/source/contributing/typescript.md:
````diff
@@ -9,14 +9,126 @@ myst:
+```diff
+--- a/Makefile
++++ b/Makefile
+@@ -59,7 +59,8 @@ preinstall: ## Preinstall task, checks if missdev (mrs-developer) is present and
````
```suggestion
@@ -59,7 +59,8 @@ preinstall: ## Preinstall task, checks if missdev (mrs-developer) is present and runs it
```

stevepiercy commented on docs/source/contributing/typescript.md:
`````diff
@@ -9,14 +9,126 @@ myst:
+```
+
+````{note}
+Do not forget to run `mrs-developer` after doing it, so the configuration gets in the right place (`tsconfig.json`) now.
`````
```suggestion
After editing your {file}`Makefile`, run `mrs-developer` with the following command, so the configuration gets in the right place ({file}`tsconfig.json`).
```

stevepiercy commented on docs/source/contributing/typescript.md:
`````diff
@@ -9,14 +9,126 @@ myst:
+
+````{note}
+Do not forget to run `mrs-developer` after doing it, so the configuration gets in the right place (`tsconfig.json`) now.
+Run
`````
```suggestion

```

stevepiercy commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+### Enable it in a project not using Volto 17 yet
 
-In order to enable TypeScript, you need to install these additional dependencies in your project:
+If you have an existing project using a Volto version previous to Volto 17 (and above 14), and you want to use TypeScript on it, you need to install these additional dependencies in your project:
```
```suggestion
To use TypeScript for an existing project in Volto versions 14.x-16.x, you need to install these additional dependencies in your project:
```

stevepiercy reviewed (changes requested):
Mostly grammar and syntax fixes, but a question or two. This is great! Thank you!

pnicolli commented on src/components/theme/TsTest/TsTest.test.tsx:
I understand why this is here in this PR, I am not sure we should keep this after the PR is ready though.

sneridagh replied:
I think it's useful to have it, in order to assess that things work. We can remove it when we have the first "real" TS code in place.

pnicolli commented on docs/source/upgrade-guide/index.md:
```diff
@@ -53,6 +53,112 @@ or use Webpack plugins, you might need to make adjustments.
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": false,
```
```suggestion
    "strict": true,
```

pnicolli commented on docs/source/contributing/typescript.md:
```diff
@@ -9,14 +9,126 @@ myst:
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": false,
```
```suggestion
    "strict": true,
```
